doc/design.md

```diff
@@ -0,0 +1,1268 @@
+
+##### Rationale
+
+Mantid currently stores uncertainties as standard deviation (except for `DataObjects::MDHistoWorkspace`).
+However, most computations require the variance for propagation of uncertainties.
+To this end it is often squared in place (breaking assumptions about, e.g., `HistogramData::Histogram` and `DataObjects::Workspace2D`, and subsequently converted back to standard deviation by taking the square root.
+This also adds a little computational overhead.
+Having a clear and well defined strategy will reduce the risk of confusion.
+
+##### Implementation
+
+The go-to tag for storing uncertainties will be `Data::Variance`.
+`Data::StdDev` will exist, but is for *access* only, i.e., attempting to create a `Variable` with that tag should fail.
+Instead `Data::StdDev` can by used in combination with `DatasetView` to *read* the standard deviations if required, e.g., for visualization purposes.
+If it turns out to be useful it would also be possible to provide setters --- as opposed to a simple assignment which is possible if the actual underlying data is referenced --- for standard deviation using `DatasetView` .
+
+
+### <a name="design-details-distribution-histogram-distinguised-by-unit"></a>"Distributions" and "histogram data" are distinguished by their unit
+
+##### Rationale
+
+Currently `API::MatrixWorkspace` (and by extension `HistogramData::Histogram`) keep a flag indicating whether data in a histogram has been divided by its bin width.
+A considerable number of algorithms are checking this manually, branching to different code sections.
+`HistogramData::Histogram` attempted to solve this by introducing getters like `Histogram::counts()` and `Histogram::frequencies()`.
+However, it seems to be useful only in a rather limited number of places, whereas it adds a significant amount of complication to the API.
+The flag adds complication and should thus be removed.
+
+##### Implementation
+
+This will require having a unit for "counts", i.e., we cannot simply use "dimensionless".
+We need to establish whether it is possible to write clean client code based on the unit-driven mechanism.
+Apart from that there should not be any major implementation difficulties.
+
+
+### <a name="design-details-no-progress-reporting-no-cancellation"></a>No progress reporting and no cancellation
+
+##### Rationale
+
+- Most operations are simple and thus fast.
+- Cancellation and progress reporting can be implemented in higher-level algorithms that *use* `Dataset`.
+- Cancellation leads to undefined side effects for in-place operation, i.e., it breaks data.
+- Adds complexity and complicates the interface.
+- `numpy` operations cannot be cancelled and do not report progress.
+
+
+### <a name="design-details-no-length-variation"></a>No support for histograms with varying length
+
+##### Rationale
+
+This has not been supported in `DataObjects::Workspace2D` for 10 years, until a very recent change, i.e., it does not appear to be strictly necessary.
```

Comment:
@soininen Yes, arbitrary masks are supported, and we should indeed consider having different "mask types", e.g., also for ROI selection. In most cases this would simply be something like `Coord::ROI` or `Coord::Mask`.

With `Dataset` being generic, it is of course up to the algorithm implementer to check for the mask. During the earlier design days I had considering making masking more "native", but I actually had trouble coming up with a way that would make it magically work "automatically" for all operations --- automatically matching masks based on the fact that they are coordinate variables is the closest to "native" that seemed like a real option.

Reply:
@soininen There is also a working example showing my current ideas regarding masking: https://github.com/mantidproject/dataset/blob/design/test/Workspace2D_test.cpp#L184.

Reply:
I especially like the idea of having a separate `Dataset` for the mask. The current implementation of `MaskWorkspace` in Mantid is confusing, to say least.

In the long run I'd very much prefer to have something like a ROI `Dataset` to supersede all those `WorkspaceIndexList`, `XMin`, `SpectraMin` etc. etc. in the algorithms.

Reply:
Good point regarding passing a ROI `Dataset` for these purposes. :+1:

Reply:
The numpy concept for this is [masked arrays](https://docs.scipy.org/doc/numpy/reference/maskedarray.html).

Reply:
Thanks for the reference, I will look into it.

Reply:
I also suggest looking at how apache arrow is handling missing data.